### PR TITLE
Add Origin header for CORS request

### PR DESCRIPTION
**Pull request by watfordjc**
When a browser makes a CORS preflight (OPTIONS) request, it sends an Origin header.

Some plugins (e.g. mod_bosh for prosody) do not consider an OPTIONS request to be a CORS request if the Origin header is missing.

This change sends an `Origin: null` header when making an OPTIONS request. While `null` could be replaced with a specific origin before compilation, the current test is looking for an `Access-Control-Allow-Origin: *` response header so null should be fine.

[CORS for Developers - 7.4. Avoid returning Access-Control-Allow-Origin: "null"](https://w3c.github.io/webappsec-cors-for-developers/#avoid-returning-access-control-allow-origin-null)

### CODE REVIEW

**Comment by iNPUTmice** [comment]
I’m not a webdev but it feels like putting `https://compliance.conversations.im` in there might be a bit cleaner. (valid url and such)

**Comment by watfordjc** [comment]
CORS is hard to get right with servers/plugins - my suggestion of `null` was based on likely edge cases and that it should be treated differently to other origins. My suggestion that the origin could be changed to non-null before compilation, given the current test is looking for `*`, is on further thought rather redundant.

Depending on the Web server, if the Origin isn't in the list of allowed origins, a 403 might be returned without any CORS headers (the case for prosody mod_bosh). Then again, some Web servers return a list of origins (which most/all browsers don't support) instead of doing something equivalent to `if (allowedOriginsList.contains($origin)) { addHeader("Access-Control-Allow-Origin: $origin") }` (also the case of prosody mod_bosh `cross_domain_bosh`, which will need a patch to fix that behaviour). I haven't checked mod_websocket's behaviour for `cross_domain_websocket`.

If GET were used instead of OPTIONS, a browser would change `Origin: https://compliance.conversations.im` to `Origin: null` if it had to follow a redirect, which is probable given the usage of HSTS headers, HSTS preload lists, and the recommended order of redirects before deploying HSTS (i.e. http://example.com -> https://example.com -> https://www.example.com).

For example:

`curl -i -H 'Origin: https://compliance.conversations.im' 'http://jabber.fu-berlin.de/.wel
l-known/host-meta'`

> HTTP/1.1 301 Moved Permanently
> ...
> Location: https://jabber.fu-berlin.de/.well-known/host-meta
> ...

`curl -i -H 'Origin: null' 'https://jabber.fu-berlin.de/.well-known/host-meta'`

> ...
> Access-Control-Allow-Origin: *
> ...

Basically, `Origin: null` should work without having to make the test follow everything in [WHATWG Fetch § 4.7. CORS-preflight fetch](https://fetch.spec.whatwg.org/#cors-preflight-fetch) (something I haven't tried to get my head around).

**Comment by moppman** [comment]
This commit makes Prosody fail the XEP-0156 test. Caas verbose output:
`check of https://prosodyhost:5281/http-bind failed. response code=200, CORS=false`
This is on current Prosody trunk, using [mod_http_altconnect](https://modules.prosody.im/mod_http_altconnect.html), as advised in the notes for the test.

**Comment by iNPUTmice** [comment]
@moppman fwiw at that stage `mod_http_altconnect` has nothing to do with it. That module only serves for discovering the endpoints. That obviously works. However it is then the actual bosh module that can’t handle the 'Origin' header.
